### ISSUE TITLE
Load new configuration level and TAG at RunTime

### ISSUE THREAD

**Issue by zulot** [tinylog-org/tinylog#304]
I use tinylog 2.5 in my configuration.

We need to change TAG, and severity log at RUNTIME, because the application is big, we can t log everything, and we just need to log something by tag for a specific request for a specified time.

I look issue #100, #102, #157, #192 and documentation https://tinylog.org/v2/extending/#custom-logging-provider

But that doesn't work in specific case.
Example, if TAG TEST is not in configuration file, and it's call, they are nothing in file. Normal.
I change and add tag TEST after and I launch reload fonction. Tinylog, doesn't log information for TEST. 
If i wait for tomorrow it's work, if i restart application it's work to.
But we need that's work at reload function already

```java
public class ReconfigurableLoggingProvider implements LoggingProvider {

        private TinylogLoggingProvider realProvider = new TinylogLoggingProvider();

        public void reload() throws InterruptedException, ReflectiveOperationException {
            realProvider.shutdown();

            Field frozen=Configuration.class.getDeclaredField("frozen");
            frozen.setAccessible(true);
            
            Method method = Configuration.class.getDeclaredMethod("load");
            method.setAccessible(true);

            frozen.setBoolean(null, false);
            Map config=(Map) method.invoke(null);
            
            Configuration.replace(config);
            
            realProvider = new TinylogLoggingProvider();
            
            frozen.setBoolean(null, true);
            method.setAccessible(false);
        }

        public ContextProvider getContextProvider() {
            return realProvider.getContextProvider();
            // dans la doc officiel 
            // a NopContextProvider is registered,
            //which is part of tinylog’s API artifact and simply ignores all registered context values
            //new NopContextProvider(); 
        }

        public Level getMinimumLevel() {
            return realProvider.getMinimumLevel();
        }

        public Level getMinimumLevel(final String tag) {
            return realProvider.getMinimumLevel(tag);
        }

        public boolean isEnabled(final int depth, final String tag, final Level level) {
            return realProvider.isEnabled(depth + 1, tag, level);
        }

        public void log(String loggerClassName, String tag, Level level, Throwable exception, MessageFormatter formatter, Object obj, Object... arguments)
        {
            realProvider.log(loggerClassName,tag,level,exception,formatter,obj,arguments);
        }

        public void log(int depth, String tag, Level level, Throwable exception, MessageFormatter formatter, Object obj, Object... arguments) {
            realProvider.log(depth + 1, tag, level, exception, formatter, obj, arguments);
        }   

        public void shutdown() throws InterruptedException {
            realProvider.shutdown();
        }
}
```

Can you help me to add this functionnality ?

**Comment by pmwmedia** [comment]
I have just created a working example project: https://github.com/tinylog-org/tinylog-reload-configuration-example

Please have a look if it solves your issues. I assume that the cause for issues is that the `TaggedLogger` instances are ignoring your tag configuration changes. If the example project or its documentation is unclear or won't solve your issues, please let me know. I will be happy to improve!

**Comment by zulot** [comment]
Thanks for your example

But Globaly that doesn't work, and it's worst than i did, with now multi bug

I explain, i start my configuration with 2 TAG : TOTO and TEST
Il try log TEST1 -> it is not write -> OK
I add TEST1, i reload, launch log -> OK
I add TEST2, i reload, launch log -> OK
I add TEST3, remove TEST1, TEST2, launch log TEST 3 -> OK
I launch TEST2 -> it is not write -> OK
I add TEST2, reload, launch log -> KO !!!
And after this point i test so much configuration.
TEST3 work every times, i add TEST10 work in start and after not. And finaly i add all tag and now il launch with TEST1 and write in log TEST3 for tag !

thx for help to use your lib

**Comment by pmwmedia** [comment]
Can you adapt https://github.com/tinylog-org/tinylog-reload-configuration-example with your configuration for reproducing the bug? Afterwards you can give me the code and I will have a look what is the cause.

**Comment by zulot** [comment]
I give you all configuration file and test.
Number is version off properties

1 - try write TAG: TEST1 -> no log -> OK
1 - try write TAG: TEST -> log -> OK
2 - after reload and del TAG TEST and ADD TEST1, try write TAG:TEST1 -> no log -> **KO**
2 - try write TAG:TEST -> no log OK
3 - I configure TAG: TOTO, TEST, TEST1, and reload configuration
3 - try write TAG:TEST1 -> no log -> **KO**
3 - try write TAG TEST -> lo -> OK

[tinylog2.zip](https://github.com/tinylog-org/tinylog/files/9811813/tinylog2.zip)

Java project :
[TinyLog2LoggingProvider.zip](https://github.com/tinylog-org/tinylog/files/9811839/TinyLog2LoggingProvider.zip)

Reload configuration is 
`public static final void reloadConfiguration(IData pipeline) throws ServiceException {
		try
		{
			ReconfigurableLoggingProvider.reload();
		}
		catch (Exception e)
		{
			// if class cast exception : means that the jar / provider was not properly loaded, see in wrapper.log 
			throw new ServiceException(e);
		}
	}`

**Comment by zulot** [comment]
The bug what i can reproduce is :  try to log with tag was not in configuration in FIRST.
After, add TAG in configuration, reload configuration, and try to log this TAG again -> result doesn't work.
I reproduce this with many TAG every time.

**Comment by pmwmedia** [comment]
```properties
## fichier utilisé pour la configuration en DEV en date du 20220929
## /!\ ne pas configurer 2 writter sur le même fichier cela produit des incohérence

provider = reconfigurable

# Register the BlackHoleWriter to ensure that disabled tags and severity levels will be not optimized away
writerTag = black hole

## Writers can be executed in a separate thread.
## The advantage is that the application itself will not be blocked by slow IO operations.
## The writing thread runs with low priority and shuts down automatically, together with the application.
#writingthread = true # optional, default: false

writerTag.format = {date:yyyy-MM-dd HH:mm:ss} - TAG : {tag} - {level} : {message}


## Shared File allows multiple instances of an application to write to the same log file simultaneously
#writerTag = shared file
writerTag = rolling file

## Séparer les différents TAG par ", " (virgule et espace)
writerTag.tag = TOTO, TEST

## valeur : trace, debug, info, warn, and error
writerTag.level = trace

## file of log example : writer.file = /{date:yyyy}/{date:MM}/log-{pid}.{count}
writerTag.file = logs/RGDSLogsSFTP-{date:yyyy-MM-dd}.log
## Optional link to the latest log file
writerTag.latest = logs/RGDSLogsSFTP.log            
writerTag.backups = 365
## deprecated ???
#writerTag.label = timestamp:yyyy-MM-dd
writerTag.policies = daily
```

You have to use different properties names for both writers.

Correct would be:

```properties
provider = reconfigurable

writerBlockHole = black hole
writerBlockHole.level = trace

writerTag = rolling file
writerTag.tag = TOTO, TEST
writerTag.level = trace
writerTag.format = {date:yyyy-MM-dd HH:mm:ss} - TAG : {tag} - {level} : {message}
writerTag.file = logs/RGDSLogsSFTP-{date:yyyy-MM-dd}.log
writerTag.latest = logs/RGDSLogsSFTP.log            
writerTag.backups = 365
writerTag.policies = daily
```

**Comment by zulot** [comment]
I don't understand.
I need only one writer. 

How writerTag can be reload with his new tag, if it's not black hole ?

I do the same test with your configuration, first i use 
writerTag.tag = TOTO, TEST
log tag : TEST1 -> no write : OK
After i use 
writerTag.tag = TOTO, TEST, TEST1
i reload
log tag : TEST1 -> no write : **KO**

**Comment by pmwmedia** [comment]
> How writerTag can be reload with his new tag, if it's not black hole ?

Block hole is a second dummy writer that does not output anything. However, it prevents any output optimizations that would make re-configuration impossible.

> I do the same test with your configuration, first i use

For me, it does work:

```
2022-10-21 20:32:24 - TAG : TEST - INFO : Attempt 1
2022-10-21 20:32:24 - TAG : TEST1 - INFO : Attempt 2
2022-10-21 20:32:24 - TAG : TEST1 - INFO : Attempt 3
2022-10-21 2ß:32:24 - TAG : TEST - INFO : Attempt 3
```
My example project: [example.zip](https://github.com/tinylog-org/tinylog/files/9841828/example.zip)

**Comment by zulot** [comment]
Hi, thx for your example,
But i reproduce bug if you use the same file in debug mod of your example.

the reload properties seems doen't work

My application : 

```java
ReconfigurableLoggingProvider.reload("/tinylog-1.properties");
Logger.tag("TEST1").info("Attempt 1");
Logger.tag("TEST").info("Attempt 1");

ReconfigurableLoggingProvider.reload("/tinylog-1.properties");
Logger.tag("TEST1").info("Attempt 2");
Logger.tag("TEST").info("Attempt 2");

ReconfigurableLoggingProvider.reload("/tinylog-1.properties");
Logger.tag("TEST1").info("Attempt 3");
Logger.tag("TEST").info("Attempt 3");`
```

I put break point in all line : ReconfigurableLoggingProvider.reload("/tinylog-1.properties");

When programm stop i change properties file and i go next step, and it's doesn't work.

result file

> 2022-10-25 16:36:24 - TAG : TEST - INFO : Attempt 1
> 2022-10-25 16:37:45 - TAG : TEST - INFO : Attempt 2
> 2022-10-25 16:38:58 - TAG : TEST - INFO : Attempt 3


After my test i use your application with 3 files it's ok
result

> 2022-10-25 16:36:24 - TAG : TEST - INFO : Attempt 1
> 2022-10-25 16:37:45 - TAG : TEST - INFO : Attempt 2
> 2022-10-25 16:38:58 - TAG : TEST - INFO : Attempt 3
> 2022-10-25 16:44:34 - TAG : TEST - INFO : Attempt 1
> 2022-10-25 16:44:34 - TAG : TEST1 - INFO : Attempt 2
> 2022-10-25 16:44:34 - TAG : TEST1 - INFO : Attempt 3
> 2022-10-25 16:44:34 - TAG : TEST - INFO : Attempt 3

One more things i see this error after this line code

`realProvider = new TinylogLoggingProvider();`

> LOGGER ERROR: Failed to create link 'logs\RGDSLogsSFTP.log' (java.nio.file.FileSystemException: logs\RGDSLogsSFTP.log: The process cannot access the file because it is being used by another process.
> )

Conclusion, that doesn't work when you want to reload the same properties file


EDIT : I'm not sur, but i extract properties, from project, i change how the properties load, and that work.
I continue my test for sur

**Comment by pmwmedia** [comment]
> When programm stop i change properties file and i go next step, and it's doesn't work.

If you want to change the tinylog configuration files during runtime, I propose to load them directly from the filesystem and not from the classpath. This should fix your issue that your file changes don't apply.

> LOGGER ERROR: Failed to create link 'logs\RGDSLogsSFTP.log' (java.nio.file.FileSystemException: logs\RGDSLogsSFTP.log: The process cannot access the file because it is being used by another process.
)

Sounds like your application is running multiple times simultaneously or the log file is opened in any editor that locks the file.

**Comment by pmwmedia** [comment]
I will close this issue for now. If there is anything else, how I can support you using tinylog, please let me know.
